### PR TITLE
mgmt: mcumgr: fs: Add checksum (CRC32), hash (SHA256) and file size handlers

### PR DESCRIPTION
**Pull request by nordicjm**
doc: release-notes: Add mcumgr file status, file hash/checksum details
    
    This adds release notes for the new file status and file hash/checksum
    mcumgr fs commands.
    
    doc: guides: device_mgmt: smp: fs: Add hash/checksum/status details
    
    This adds documentation which explains fs mcumgr commands for the status
    and hash/checksum commands.
    
    mgmt: mcumgr: fs: Add hash/checksum (with CRC32/SHA256) handler
    
    This adds a hash/checksum mcumgr handler to the file management commands
    which can be used to get a hask or checksum of a file, and includes
    handler implementations for IEEE CRC32 and SHA256.
    
    mgmt: mcumgr: fs: Add file status handler
    
    This adds a command handler to the file management mcumgr system to get
    the status of a file without needing to return file data (currently
    reporting the file size).

### CODE REVIEW

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include/fs_mgmt/fs_mgmt.h [hunk 1]
```diff
@@ -14,7 +14,9 @@ extern "C" {
 /**
  * Command IDs for file system management group.
  */
-#define FS_MGMT_ID_FILE	 0
+#define FS_MGMT_ID_FILE		0
+#define FS_MGMT_ID_CRC32	1
+#define FS_MGMT_ID_SHA256	2
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include/fs_mgmt/fs_mgmt.h:19]: Do not add these as separate commands, use one command like `FS_MGMT_ID_CHKSUM` instead with param that would allow to select checksum algorithm. Otherwise we start to bloat IDs with all the requested checksums.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 2]
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+	rc = fs_mgmt_impl_checksum(path, (enum fs_mgmt_checksum_t)type,
+				   &checksum, &file_len, &checksum_len);
+
+	if (rc != 0) {
+		return rc;
+	}
+
+	/* Encode the response. */
+	err = 0;
+	err |= cbor_encode_text_stringz(&ctxt->encoder, "rc");
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:314]: I see no point in assigning here 0, just assign result of first cbor call directly.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 3]
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+		{ 0 },
+	};
+
+	type = FS_MGMT_CHECKSUM_COUNT;
+	rc = cbor_read_object(&ctxt->it, size_attr);
+
+	if (rc != 0 || type >= FS_MGMT_HASH_COUNT) {
+		return MGMT_ERR_EINVAL;
+	}
+
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:371]: I think that it is worth adding `path[ARRAY_SIZE(path) - 1] = '\0'` here to "just-in-case" make sure that null terminator is there.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 4]
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+		{ 0 },
+	};
+
+	type = FS_MGMT_CHECKSUM_COUNT;
+	rc = cbor_read_object(&ctxt->it, size_attr);
+
+	if (rc != 0 || type >= FS_MGMT_CHECKSUM_COUNT) {
+		return MGMT_ERR_EINVAL;
+	}
+
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:303]: I think that it is worth adding `path[ARRAY_SIZE(path) - 1] = '\0'` here to "just-in-case" make sure that null terminator is there.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 5]
```diff
@@ -214,11 +215,205 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+		.mh_write = NULL,
+	},
+	[FS_MGMT_ID_CHECKSUM] = {
+		.mh_read = fs_mgmt_file_checksum,
+		.mh_write = NULL,
+	},
+	[FS_MGMT_ID_HASH] = {
+		.mh_read = fs_mgmt_file_hash,
+		.mh_write = NULL,
+	},
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:416]: When they are not supported, do not add them here: the mcumgr will respond with error 8 when unable to find handler, and adding these makes it just do sequence:
1) search handler
2) call handler
3) handler returns with error
4) reset SMP buffer and form "rc" response with error returned from handler

Instead of directly going:
1) search handler;
1) send "rc" response with error 8
Additionally this adds code, when commands are not supported, that duplicates already existing behavior of SMP request processor.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/zephyr_fs_mgmt.c [hunk 6]
```diff
@@ -28,9 +42,7 @@ fs_mgmt_impl_filelen(const char *path, size_t *out_len)
 	return 0;
 }
 
-int
-fs_mgmt_impl_read(const char *path, size_t offset, size_t len,
-				  void *out_data, size_t *out_len)
+int fs_mgmt_impl_read(const char *path, size_t offset, size_t len, void *out_data, size_t *out_len)
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/zephyr_fs_mgmt.c:45]: Style changes like this, the `zephyr_fs_mgmt_truncate` and `zephyr_fs_mgmt_truncate` should at least get separate commit, or PR. Otherwise this just bloats PR diff with changes that are not related to the subject of the change.

**Review by de-nordic — CHANGES_REQUESTED** [review]
Aside from comments, also needed:
1) release notes entry on new Kconfigs and commands added to mcumgr;
2) SMP documentation update, probably - new commands are added.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/zephyr_fs_mgmt.c (hunk 6 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/zephyr_fs_mgmt.c:45]: This was caused by using clang format to try and get rid of the 'do...while' error, should I just revert the changes and ignore that error?

**Comment by de-nordic** [comment]
> @de-nordic Does it make sense to have an additional function which returns an array/map of the supported checksum/hash functions on a device?

No and yes, it depends: In most cases you would know what sums you have implemented and what you want to request, on the other side you may have some devices that have newer software and support additional sums and you may query the device to know what sums you may use.

Log mgmt/mcumgr/lib Feature request for this so we can focus on having the current PR merged.

**Comment by lairdjm** [comment]
Added length and offset parameters

**Review by nvlsianpu — COMMENTED** [review]
@de-nordic As I understand you are accepting addition of functionalities in this PR. Am I right?

File: doc/guides/device_mgmt/smp_groups/smp_group_8.rst [hunk 7]
```diff
@@ -216,3 +220,209 @@ where:
...
+    | ``1``  | ``8``        |  ``1``         |
+    +--------+--------------+----------------+
+
+CBOR data of successful response:
+
+.. code-block:: none
+
+    {
+        (str)"len"      : (uint)
+        (str)"rc"       : (int)
```
Comment (de-nordic) [inline doc/guides/device_mgmt/smp_groups/smp_group_8.rst:281]: Do the rc optional; the rc should only be sent when there is an error.

File: doc/guides/device_mgmt/smp_groups/smp_group_8.rst [hunk 8]
```diff
@@ -216,3 +220,209 @@ where:
...
+    +========+==============+================+
+    | ``0``  | ``8``        |  ``2``         |
+    +--------+--------------+----------------+
+
+CBOR data of request:
+
+.. code-block:: none
+
+    {
+        (str)"type"     : (str)
```
Comment (de-nordic) [inline doc/guides/device_mgmt/smp_groups/smp_group_8.rst:334]: Make type optional. The type is returned in response anyway and very often users will only rely on whatever default type would be set.

File: subsys/mgmt/mcumgr/Kconfig [hunk 9]
```diff
@@ -131,6 +131,56 @@ config FS_MGMT_DL_CHUNK_SIZE
 
 endif
 
+config FS_MGMT_FILE_STATUS_ENABLED
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/Kconfig:134]: `_ENABLED` non needed. For example look at this `IS_ENABLED(CONFIG_FS_MGMT_FILE_STATUS_ENABLED)` would look strange.

File: subsys/mgmt/mcumgr/Kconfig [hunk 10]
```diff
@@ -131,6 +131,56 @@ config FS_MGMT_DL_CHUNK_SIZE
...
+	bool "File status command enabled"
+	default y
+	help
+	  This command allows a remote device to retrieve the status of a file,
+	  at present only the size of the file is returned (if it exists).
+
+config FS_MGMT_CHECKSUM_HASH_SUPPORTED
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/Kconfig:141]: Just `FS_MGMT_CHECKSUM_HASH`

File: subsys/mgmt/mcumgr/Kconfig [hunk 11]
```diff
@@ -131,6 +131,56 @@ config FS_MGMT_DL_CHUNK_SIZE
...
+	bool "Checksum/hash mcumgr functions"
+	depends on HEAP_MEM_POOL_SIZE != 0
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/Kconfig:143]: This is very weak dependency, you can set the `HEAP_MEM_POOL_SIZE` to any non-0 and the functions will still not work because something else will take all of the heap.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 12]
```diff
@@ -208,11 +226,235 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+		return MGMT_ERR_ENOENT;
+	}
+
+	if (file_len <= off) {
+		/* Requested offset is larger than target file size */
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Allocate a buffer for the output data */
+	output = k_malloc(group->output_size);
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:355]: I do not think that it is worth to engage heap for this. Just define buffer that will take the longest currently supported hash, as selected by Kconfig, either on stack or as static buffer.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 13]
```diff
@@ -208,11 +226,235 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+	}
+
+	/* Check provided offset is valid for target file */
+	rc = fs_mgmt_impl_filelen(path, &file_len);
+
+	if (rc != 0) {
+		return MGMT_ERR_ENOENT;
+	}
+
+	if (file_len <= off) {
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:349]: Shouldn't this be
```
if (file_len < off + len)
```
?

Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:349]: `file_len` is the maximum size to use, so if the file size is not known then the maximum value would be accepted rather than having to query the file size first and then getting a hash/checksum.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+		if ((read_size + *out_len) >= len) {
+			/* Limit read size to size of requested data */
+			read_size = len - *out_len;
+		}
+
+		bytes_read = fs_read(file, buffer, read_size);
+		crc32 = crc32_ieee_update(crc32, buffer, bytes_read);
+		*out_len += bytes_read;
+
+		if (*out_len >= len) {
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c:40]: This can be incorporated into `while` loop condition.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+			/* All of the requested file data has been processed */
+			break;
+		}
+	} while (bytes_read > 0);
+
+	memcpy(output, &crc32, sizeof(crc32));
+
+	return 0;
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c:48]: Shouldn't we return error here when `bytes_read < 0`? User will only notice failure when it is know what is file size, otherwise this just returns crc for any number of bytes read and that size will be reported in response.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 16]
```diff
@@ -208,11 +226,235 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+
+	/* Calculate hash/checksum using function */
+	file_len = 0;
+	rc = group->function(ctxt, &file, output, &file_len, len);
+
+	fs_close(&file);
+
+	/* Encode the response */
+	err = cbor_encode_text_stringz(&ctxt->encoder, "rc");
+	err |= cbor_encode_int(&ctxt->encoder, rc);
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:389]: No, don't. We should stop using "rc" for reporting non SMP errors and success does not OK to be sent back.

**Comment by de-nordic** [comment]
> @de-nordic As I understand you are accepting addition of functionalities in this PR. Am I right?

Yes. We have discussed them and I find them to be nice addition.

File: doc/guides/device_mgmt/smp_groups/smp_group_8.rst (hunk 8 above)
Comment (lairdjm) [reply doc/guides/device_mgmt/smp_groups/smp_group_8.rst]: Done, made type selection automatic however.

File: subsys/mgmt/mcumgr/Kconfig (hunk 9 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/Kconfig]: Done

File: subsys/mgmt/mcumgr/Kconfig (hunk 10 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/Kconfig]: Done

File: subsys/mgmt/mcumgr/Kconfig (hunk 11 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/Kconfig]: Changed to stack, default at 128

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c (hunk 12 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c]: Uses longest supported hash/checksum size for buffer

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_sha256.c [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+	if (mbedtls_md_setup(&mbed_hash_ctx, mbed_hash_info, 0) != 0) {
+		return MGMT_ERR_EUNKNOWN;
+	}
+
+	if (mbedtls_md_starts(&mbed_hash_ctx)) {
+		rc = MGMT_ERR_EUNKNOWN;
+		goto error;
+	}
+#endif
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_sha256.c]: Done

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c (hunk 14 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c]: Done

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c (hunk 16 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c]: Done

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c (hunk 15 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c:49]: I've added an error in but just realised it's a zephyr errno code rather than mgmt error code, should I do a mapping to a mgmt error code or leave it as an errno code and update the documentation?

Comment (de-nordic) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c:49]: The SMP and errno should not overlap as SMP errors are positive and errnos are usually returned as negative.

Anywa,  I have proposed https://github.com/zephyrproject-rtos/zephyr/issues/42590 where I state my doubts whether "rc" should be used for errors not related to SMP processing, and here an error would not be related to it - because we have been able to process the SMP but failed at executing some operations. 
I was wandering whether in case of calculation/read/internal error the "rc", from SMP, should still be zero, but instead additional "err" code is logged where the system error is given.

Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_crc32.c:49]: Fixed to use UNKNOWN error for now

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 18]
```diff
@@ -1,17 +1,43 @@
...
+
+#if defined(CONFIG_FS_MGMT_HASH_SHA256)
+#include "fs_mgmt/hash_checksum_sha256.h"
+#endif
+
+#ifdef CONFIG_FS_MGMT_CHECKSUM_HASH
+BUILD_ASSERT(sizeof(CONFIG_FS_MGMT_CHECKSUM_HASH_DEFAULT) > 1,
+	     "Missing mcumgr fs checksum/hash algorithm selection?");
+BUILD_ASSERT(CONFIG_FS_MGMT_CHECKSUM_HASH_LARGEST_OUTPUT_SIZE > 0,
+	     "Missing mcumgr fs checksum/hash algorithm selection?");
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:33]: I am not sure whether this is needed; an user will not be able to set either of these with proj.conf because it is not allowed to set Kconfig options without prompt.

Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:33]: They can enable the feature and not enable any algorithms for it which this catches

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 19]
```diff
@@ -208,11 +234,227 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+				tmp_val = (uint64_t)(*(uint8_t *)output);
+			} else if (group->output_size == sizeof(uint16_t)) {
+				tmp_val = (uint64_t)(*(uint16_t *)output);
+			} else if (group->output_size == sizeof(uint32_t)) {
+				tmp_val = (uint64_t)(*(uint32_t *)output);
+			} else if (group->output_size == sizeof(uint64_t)) {
+				tmp_val = (*(uint64_t *)output);
+			} else {
+				LOG_ERR("Unable to handle numerical checksum size %d",
+					group->output_size);
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c]: Fixed

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+
+	/* Find the group with the specified group name */
+	for (group = hash_checksum_mgmt_group_list; group != NULL; group = group->hc_mg_next) {
+		if (strcmp(group->group_name, name) == 0) {
+			break;
+		}
+	}
+
+	return group;
+}
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:56]: Merge with `hash_checksum_mgmt_find_handler`, it seems that currently `hash_checksum_mgmt_find_handler` is only null check on return value from calling this.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+	hash_checksum_mgmt_group_list_end = group;
+}
+
+const struct hash_checksum_mgmt_group *hash_checksum_mgmt_find_handler(const char *name)
+{
+	const struct hash_checksum_mgmt_group *group;
+
+	group = hash_checksum_mgmt_find_group(name);
+	if (!group) {
+		return NULL;
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:74]: I think that `return group` at line 77 is enough here.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	if (!prev || !curr) {
+		return;
+	}
+
+	prev->hc_mg_next = curr->hc_mg_next;
+	if (curr->hc_mg_next == NULL) {
+		hash_checksum_mgmt_group_list_end = curr;
+	}
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:41]: You may use slist from Zephyr (https://docs.zephyrproject.org/latest/kernel/data_structures/slist.html) here.

**Review by de-nordic — CHANGES_REQUESTED** [review]

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c [hunk 23]
```diff
@@ -187,11 +213,217 @@ fs_mgmt_file_upload(struct mgmt_ctxt *ctxt)
...
+
+	fs_close(&file);
+
+	/* Encode the response */
+	if (rc != 0) {
+		ok = zcbor_tstr_put_lit(zse, "rc")	&&
+		     zcbor_int32_put(zse, rc);
+	} else {
+		ok = zcbor_tstr_put_lit(zse, "type")	&&
+		     zcbor_tstr_put_term(zse, type_arr);
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:365]: `type` can only be used if the type was supplied, otherwise it will be empty as `type_arr` contains the default, unless you mean to do an if for which function to call and variable to use here?

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/Kconfig [hunk 24]
```diff
@@ -89,11 +89,65 @@ config FS_MGMT_DL_CHUNK_SIZE
...
+	default "sha256" if FS_MGMT_HASH_SHA256
+	help
+	  Default checksum/hash type if not specified.
+
+config FS_MGMT_CHECKSUM_HASH_LARGEST_OUTPUT_SIZE
+	int
+	default 32 if FS_MGMT_HASH_SHA256
+	default 4 if FS_MGMT_CHECKSUM_IEEE_CRC32
+	help
+	  Largest checksum/hash output size for internal buffer allocation.
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/Kconfig]: Fixed

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include/fs_mgmt/hash_checksum_mgmt.h [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+#include <fs/fs.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @typedef hash_checksum_mgmt_handler_fn
+ * @brief Function that gets called to generate a hash or checksum.
+ *
+ * @param ctxt		mcumgr mgmt context
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/include/fs_mgmt/hash_checksum_mgmt.h]: Fixed (was carry over from original PR where the hash function added data directly)

File: doc/services/device_mgmt/smp_groups/smp_group_8.rst [hunk 26]
```diff
@@ -216,3 +220,215 @@ where:
...
+
+    +-----------------------+---------------------------------------------------+
+    | "len"                 | length of file (in bytes)                         |
+    +-----------------------+---------------------------------------------------+
+    | "rc"                  | :ref:`mcumgr_smp_protocol_status_codes`           |
+    +-----------------------+---------------------------------------------------+
+
+In case when "rc" is not 0, success, the other fields will not appear.
+
+fuck
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply doc/services/device_mgmt/smp_groups/smp_group_8.rst]: Good reminder to remove rc from the documentation, fixed.

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+static struct hash_checksum_mgmt_group *hash_checksum_mgmt_group_list;
+static struct hash_checksum_mgmt_group *hash_checksum_mgmt_group_list_end;
+
+void hash_checksum_mgmt_unregister_group(struct hash_checksum_mgmt_group *group)
+{
+	struct hash_checksum_mgmt_group *curr = hash_checksum_mgmt_group_list, *prev = NULL;
+
+	if (!group) {
+		return;
+	}
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c]: Fixed

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c (hunk 20 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c]: Fixed

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c (hunk 21 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c]: Fixed

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c (hunk 22 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c:42]: This the base code was a copy and paste from the mgmt handler functions, guess we need a bug to update that also

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (de-nordic) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/fs_mgmt.c:377]: Right, I missed that. Leave it as is.

File: doc/releases/release-notes-3.1.rst [hunk 28]
```diff
@@ -187,6 +187,12 @@ Build and Infrastructure
 Libraries / Subsystems
 **********************
 
+* Management
+
+  * Added mcumgr fs handler for getting file status which returns file size
+  * Added mcumgr fs handler for getting file hash/checksum, with support for
+    IEEE CRC32 and SHA256
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply doc/releases/release-notes-3.1.rst]: Updated

File: doc/releases/release-notes-3.1.rst [hunk 29]
```diff
@@ -187,6 +187,12 @@ Build and Infrastructure
 Libraries / Subsystems
 **********************
 
+* Management
+
+  * Added mcumgr fs handler for getting file status which returns file size
```
[the remark this answers is not in the document]
Comment (lairdjm) [reply doc/releases/release-notes-3.1.rst]: Updated

File: subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c (hunk 22 above)
Comment (lairdjm) [reply subsys/mgmt/mcumgr/lib/cmd/fs_mgmt/src/hash_checksum_mgmt.c]: FIxed

**Review by de-nordic — APPROVED** [review]
Looks OK. Thanks!

**Review by carlescufi — APPROVED** [review]
